speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/frontend/IndicatorConfig.cpp

```diff
@@ -0,0 +1,164 @@
+/// @return         The pointer to the struct containing the sound data
+tSoundData* IndicatorConfig::LoadSound(void* p_handle, std::string p_path)
+{
+    p_path += PRM_SECT_SOUND;
+    if (!GfParmExistsSection(p_handle, p_path.c_str())) return nullptr;
+
+    tSoundData* data = new SoundData;
+
+    const char* source = GfParmGetStr(p_handle, p_path.c_str(), PRM_ATTR_SRC, "");
+
+    char* sndPath = new char[PATH_BUF_SIZE];
+    snprintf(sndPath, PATH_BUF_SIZE, SOUNDS_DIR_FORMAT, GfDataDir(), source);
+    data->Path = sndPath;
+
+    return data;
+}
+
+/// @brief          Loads the screen position data from the indicator config.xml
+/// @param p_handle The p_handle to the config.xml file
+/// @param p_path   The C-string pointer to the section containing x- and y-pos attributes
+/// @return         A struct containing the screen position data
+tScreenPosition IndicatorConfig::LoadScreenPos(void* p_handle, const char* p_path)
+{
+    float xPos = GfParmGetNum(p_handle, p_path, PRM_ATTR_XPOS, NULL, 0);
+    float yPos = GfParmGetNum(p_handle, p_path, PRM_ATTR_YPOS, NULL, 0);
+
+    // Check whether x- and y-pos are valid percentages in range [0,1]
+    if (xPos < 0.0f || yPos < 0.0f || xPos > 1.0f || yPos > 1.0f)
+    {
+        throw std::out_of_range("X and Y positions should be in the range [0,1]");
+    }
+    return { xPos, yPos };
+}
+
+/// @brief          Loads the texture indicator data from the indicator config.xml
+/// @param p_handle The p_handle to the config.xml file
+/// @param p_path   The p_path to the current intervention action to load
+/// @return         The pointer to struct containing the texture data
+tTextureData* IndicatorConfig::LoadTexture(void* p_handle, std::string p_path)
+{
+    p_path += PRM_SECT_TEXTURE;
+    if (!GfParmExistsSection(p_handle, p_path.c_str())) return nullptr;
+
+    tTextureData* data = new TextureData;
+    data->Path = GfParmGetStr(p_handle, p_path.c_str(), PRM_ATTR_SRC, "");
+    data->ScrPos = LoadScreenPos(p_handle, p_path.c_str());
+    return data;
+}
+
+/// @brief          Loads the text indicator data from the indicator config.xml
+/// @param p_handle The p_handle to the config.xml file
+/// @param p_path   The p_path to the current intervention action to load
+/// @return         The pointer to struct containing the text data
+tTextData* IndicatorConfig::LoadText(void* p_handle, std::string p_path)
+{
+    p_path += PRM_SECT_TEXT;
+    if (!GfParmExistsSection(p_handle, p_path.c_str())) return nullptr;
+
+    tTextData* data = new TextData;
+    data->Text = GfParmGetStr(p_handle, p_path.c_str(), PRM_ATTR_CONTENT, "");
+    data->ScrPos = LoadScreenPos(p_handle, p_path.c_str());
+    return data;
+}
+
+/// @brief  Returns an the existing IndicatorConfig, or creates the initial one if it doesn't exist yet.
+/// @return The IndicatorConfig instance
+IndicatorConfig* IndicatorConfig::GetInstance() 
+{
+    if (m_instance == nullptr)
```

Comment:
Instead of if it does not exist make it, otherwise return it. It is better the other way around: if it does exist return it, otherwise make it. So: 
if (m_instance != nullptr) return m_instance; 
or even shorter: 
if (m_instance) return m_instance;

Reply:
This was copied from the Mediator singleton, but I will update both. 
@KungFuDonkey I updated the singleton Mediator to fit the same style.